### PR TITLE
libarchive: don't attempt to make streams for nonexistent files

### PR DESCRIPTION
**Pull request by Dudemanguy911**
If built with libarchive support, mpv will use libarchive to open rar
files instead of its own rar demuxer. stream_libarchive checks if it is
a multi-part rar with some string/pattern matching and then attempts at
making a stream. This works but outputs an annoying error message that
can potentially be confusing. Avoid this by checking if the file even
exists first before attempting to make a stream.

Disclaimer: I haven't actually tested a multi-part rar (don't have one and the `rar` package on my distro doesn't seem to work right), but this shouldn't alter whatever the behavior was before.

I agree that my changes can be relicensed to LGPL 2.1 or later.

### CODE REVIEW

File: stream/stream_libarchive.c [hunk 1]
```diff
@@ -207,6 +207,12 @@ static char **find_volumes(struct stream *primary_stream)
     struct bstr base = bstr_splice(primary_url, 0, -strlen(pattern->match));
     for (int i = pattern->start; i <= pattern->stop; i++) {
         char* url = pattern->volume_url(res, pattern->format, base, i);
+        FILE *f = fopen(url, "r");
```
Comment (sfan5) [inline stream/stream_libarchive.c]: [access(2)](https://linux.die.net/man/2/access)

Comment (Dudemanguy911) [reply stream/stream_libarchive.c]: That's definitely smarter.

**Comment by sfan5** [comment]
The problem I see is that this would break opening multi-part rar files that aren't local e.g. over HTTP.

**Comment by Dudemanguy911** [comment]
Good catch. I'll need to think of a smarter solution.

**Comment by Dudemanguy911** [comment]
Alright, I believe this should do it now. ffmpeg has an [avio_check](https://ffmpeg.org/doxygen/3.0/avio_8h.html#aa4006f7b803e4d104231734a23b35ab7) function we can use to check whether a resource can be accessed. I've tested it on this [example](https://www.philippwinterberg.com/download/example_split_archive.part1.rar) multi-part rar online and it is accessed correctly. It won't actually display because there's a pdf inside, but the url and its subsequent parts are all found.

**Comment by wnoun** [comment]
`avio_check` cann't work with `stream_info_t`, it will break the `stream_cb` api.

**Comment by Dudemanguy911** [comment]
This doesn't touch `stream_info_t` or any of the stream-related functions at all though. `mp_archive_new` isn't a part of `stream_cb` and is specifically just for the libarchive demuxer, so I don't understand how it could possibly be effected. `avio_check` was chosen because it turns out to be a convenient way to check if something exists over a network or locally.

**Comment by wnoun** [comment]
becuase `stream_info_libarchive` itself doesn't open the file. For example, if you open a `smb://` archive, the `stream_info_smb` is the upstream of `stream_info_libarchive`. `avio_check` isn't able to handle all protocols.

File: stream/stream_libarchive.c [hunk 2]
```diff
@@ -207,6 +208,11 @@ static char **find_volumes(struct stream *primary_stream)
     struct bstr base = bstr_splice(primary_url, 0, -strlen(pattern->match));
     for (int i = pattern->start; i <= pattern->stop; i++) {
         char* url = pattern->volume_url(res, pattern->format, base, i);
+        int not_found = avio_check(url, 0);
+        if (not_found) {
+            talloc_free(url);
+            goto done;
+        }
         struct stream *s = stream_create(url, STREAM_READ | STREAM_SAFE_ONLY,
```
Comment (wnoun) [inline stream/stream_libarchive.c]: In more detail, `stream_create` here is the mpv version of `avio_check`

**Comment by Dudemanguy911** [comment]
Oh I understand what you mean now. Passing it through a protocol that exists in mpv but not ffmpeg (libsmbclient is the only one I can think of at the moment but perhaps there's more) will fail the check. Fortunately, avio_check is smart enough to realize if a protocol is unsupported and returns a different error code in that case. So instead, I'll just make this check more strict and the condition is only triggered if ffmpeg returns specifically that no file/directory was found. So if you open a rar with `smb://`, you will still get the error message, but not for the vast majority of use cases. So it's still a net improvement.

It's worth noting that the rar demuxer doesn't have this problem thanks to its RarProbe function.

**Comment by Dudemanguy911** [comment]
Yeah I'd have to think of some other way to hack around the warning.

**Comment by Dudemanguy911** [comment]
Can't say I really have any interest in this one anymore, so I might as well close it.

**Pull request closed without merging**